### PR TITLE
80 reuse rendered video to multiple consumers

### PR DESCRIPTION
**Pull request by yex33**

### CODE REVIEW

**Review by jiaoa3 — APPROVED** [review]

File: src/back/server.py [hunk 1]
```diff
@@ -6,171 +6,133 @@
...
 
 async def consumer(request):
     if request.method == "OPTIONS":
         return web.Response(
             content_type="application/json",
-            headers={"Access-Control-Allow-Origin": "*", 
-                        "Access-Control-Allow-Credentials": "true", 
-                        "Access-Control-Allow-Methods": "POST, GET, OPTIONS", 
-                        "Access-Control-Allow-Headers": "Content-Type"},
+            headers={
```
Comment (qstommyshu) [inline src/back/server.py:63]: All the response headers are the same, they can be used as a constant

File: src/back/server.py [hunk 2]
```diff
@@ -6,171 +6,133 @@
...
-            # Rebuild a VideoFrame, preserving timing information
-            new_frame = VideoFrame.from_ndarray(img, format="bgr24")
-            new_frame.pts = frame.pts
-            new_frame.time_base = frame.time_base
-            return new_frame
-        else:
-            return frame
+source_video = VideoStreamTrack()
+NO_TRANSFORM: str = "none"
+__transformed_tracks = dict()
```
Comment (qstommyshu) [inline src/back/server.py:28]: This is not a class attribute, why the variable starts with two underscore?

File: src/back/server.py [hunk 3]
```diff
@@ -6,171 +6,133 @@
...
+    except ValidationError as error:
+        raise web.HTTPBadRequest(reason=error.message)
 
+    description = RTCSessionDescription(sdp=body["sdp"], type=body["type"])
     pc = RTCPeerConnection()
     pc_id = "PeerConnection(%s)" % uuid.uuid4()
-    pcs.add(pc)
+    peer_connections.add(pc)
 
     def log_info(msg, *args):
```
Comment (qstommyshu) [inline src/back/server.py:82]: please don't interleave function definition with other code logic

File: src/back/server.py [hunk 4]
```diff
@@ -6,171 +6,133 @@
...
+            "Access-Control-Allow-Credentials": "true",
+            "Access-Control-Allow-Methods": "POST, GET, OPTIONS",
+            "Access-Control-Allow-Headers": "Content-Type"
+        },
     )
 
 
 async def broadcast(request):
     if request.method == "OPTIONS":
         return web.Response(
```
Comment (qstommyshu) [inline src/back/server.py:115]: both broadcaster and consumer web response are in the same format, put them into  a function to reduce code duplication.

File: src/back/server.py [hunk 5]
```diff
@@ -6,171 +6,133 @@
...
+    "type": "object",
+    "properties": {
+        "sdp": {"type": "string"},
+        "type": {"type": "string"},
+    },
+    "required": ["sdp", "type"],
+}
+
 
 async def consumer(request):
```
Comment (qstommyshu) [inline src/back/server.py:59]: Please name the broadcast and consume function in a consistent manner, either "broadcast" vs "consume" or "broadcaster" vs "consumer"

File: src/back/server.py [hunk 6]
```diff
@@ -277,6 +220,4 @@ async def process_frame_for_skeleton(frame):
     app.router.add_options("/broadcast", broadcast)
     app.router.add_post("/consumer", consumer)
```
Comment (qstommyshu) [inline src/back/server.py:221]: please name the broadcast endpoint and the consume endpoint in a consistent manner

**Review by qstommyshu — CHANGES_REQUESTED** [review]

File: src/back/server.py (hunk 4 above)
Comment (yex33) [reply src/back/server.py:115]: Wont fix: after introducing a constant for the header, we're just saving two lines here inside a return statement

File: src/back/server.py (hunk 3 above)
Comment (yex33) [reply src/back/server.py:82]: Wont fix: this is a closure for more convenient logging. And this is from the aiortc example code.

File: src/back/server.py (hunk 6 above)
Comment (yex33) [reply src/back/server.py:221]: will resolve in another issue because this require both frontend and backend changes

**Review by qstommyshu — APPROVED** [review]

**Pull request merged**